### PR TITLE
RavenDB-21521: Fix for memory corruption at CompactKey

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21521

### Additional description
The suspicion that the corruption is caused by compact key memory handling lead us to disable the original mechanism that introduce it which was optimizing access to the compact keys during lock convoy critical code path. This PR implements the fix for the memory condition. It is important to note that the error was already present, it was just harder for it to trigger with a single `ArrayPool` because of the sharing dynamics. 

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [x] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

**Comment by redknightlois** [comment]
Re-enabled the optimization and fixed the `CompactKey` issue causing the corruption.

File: test/FastTests/Voron/CompactTrees/RavenDB-21521.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                _currentIdx++;
+
+                result = output.AsSpan();
+                return true;
+            }
+        }
+
+        [RavenTheory(RavenTestCategory.Voron)]
+        [InlineDataWithRandomSeed]
+        public void CompactKeyFuzzy(int seed)
```
Comment (maciejaszyk) [inline test/FastTests/Voron/CompactTrees/RavenDB-21521.cs:72]: Has this happened with every seed? Maybe it's worth adding seeds that have reproduced the problem?

Comment (arekpalinski) [reply test/FastTests/Voron/CompactTrees/RavenDB-21521.cs:72]: I agree, we also need to have test cases with the seeds for which it failed

Comment (redknightlois) [reply test/FastTests/Voron/CompactTrees/RavenDB-21521.cs:72]: All of them reproduce it :D .. after figuring out how to trigger it was the most consistent fuzzy test I have seen. It wouldnt pass even if by mistake.

**Review by arekpalinski — APPROVED** [review]

**Review by maciejaszyk — APPROVED** [review]

**Comment by arekpalinski** [comment]
test this in debug please

**Pull request merged**